packages/EDA/src/pls/pls-tools.ts

```diff
@@ -0,0 +1,389 @@
+// Tools for multivariate analysis by PLS
+
+import * as grok from 'datagrok-api/grok';
+import * as ui from 'datagrok-api/ui';
+import * as DG from 'datagrok-api/dg';
+
+import {PLS_ANALYSIS, ERROR_MSG, TITLE, HINT, LINK, COMPONENTS, INT, TIMEOUT,
+  RESULT_NAMES, WASM_OUTPUT_IDX, RADIUS, LINE_WIDTH, COLOR, X_COORD, Y_COORD,
+  DEMO_INTRO_MD, DEMO_RESULTS_MD, DELAY, DEMO_RESULTS} from './pls-constants';
+import {checkWasmDimensionReducerInputs, checkColumnType, checkMissingVals} from '../utils';
+import {_partialLeastSquareRegressionInWebWorker} from '../../wasm/EDAAPI';
+import {carsDataframe} from '../data-generators';
+
+const min = Math.min;
+const max = Math.max;
+
+/** PLS analysis results */
+export type PlsOutput = {
+  prediction: DG.Column<DG.COLUMN_TYPE.FLOAT>,
+  regressionCoefficients: DG.Column<DG.COLUMN_TYPE.FLOAT>,
+  tScores: DG.Column<DG.COLUMN_TYPE.FLOAT>[],
+  uScores: DG.Column<DG.COLUMN_TYPE.FLOAT>[],
+  xLoadings: DG.Column<DG.COLUMN_TYPE.FLOAT>[],
+  yLoadings: DG.Column<DG.COLUMN_TYPE.FLOAT>,
+};
+
+/** PLS analysis input */
+export type PlsInput = {
+  table: DG.DataFrame,
+  features: DG.ColumnList,
+  predict: DG.Column,
+  components: number,
+  names : DG.Column | null,
+};
+
+/** Partial least square regression (PLS) */
+export async function getPlsAnalysis(input: PlsInput): Promise<PlsOutput> {
+  checkWasmDimensionReducerInputs(input.features, input.components);
+
+  // Check the responce column
+  checkColumnType(input.predict);
+  checkMissingVals(input.predict);
+
+  const result = await _partialLeastSquareRegressionInWebWorker(
+    input.table,
+    input.features,
+    input.predict,
+    input.components,
+  );
+
+  return {
+    prediction: result[WASM_OUTPUT_IDX.PREDICTION],
+    regressionCoefficients: result[WASM_OUTPUT_IDX.REGR_COEFFS],
+    tScores: result[WASM_OUTPUT_IDX.T_SCORES],
+    uScores: result[WASM_OUTPUT_IDX.U_SCORES],
+    xLoadings: result[WASM_OUTPUT_IDX.X_LOADINGS],
+    yLoadings: result[WASM_OUTPUT_IDX.Y_LOADINGS],
+  };
+}
+
+/** Perform multivariate analysis using the PLS regression */
+async function performMVA(input: PlsInput, analysisType: PLS_ANALYSIS): Promise<void> {
```

Comment:
Not sure if I like the `analysisType: PLS_ANALYSIS` idea from the architecture / function decomposition point of view - in my experience, it's easier to have three separate functions. But it's up to you :)

Reply:
Let's use a single function